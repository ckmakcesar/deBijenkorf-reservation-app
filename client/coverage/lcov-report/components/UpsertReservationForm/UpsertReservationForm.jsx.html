
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/UpsertReservationForm/UpsertReservationForm.jsx</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> / <a href="index.html">components/UpsertReservationForm</a> UpsertReservationForm.jsx</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">40% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>12/30</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">28.57% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>8/28</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">33.33% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>2/6</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">40.74% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>11/27</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <pre><table class="coverage">
<tr><td class="line-count quiet"><a name='L1'></a><a href='#L1'>1</a>
<a name='L2'></a><a href='#L2'>2</a>
<a name='L3'></a><a href='#L3'>3</a>
<a name='L4'></a><a href='#L4'>4</a>
<a name='L5'></a><a href='#L5'>5</a>
<a name='L6'></a><a href='#L6'>6</a>
<a name='L7'></a><a href='#L7'>7</a>
<a name='L8'></a><a href='#L8'>8</a>
<a name='L9'></a><a href='#L9'>9</a>
<a name='L10'></a><a href='#L10'>10</a>
<a name='L11'></a><a href='#L11'>11</a>
<a name='L12'></a><a href='#L12'>12</a>
<a name='L13'></a><a href='#L13'>13</a>
<a name='L14'></a><a href='#L14'>14</a>
<a name='L15'></a><a href='#L15'>15</a>
<a name='L16'></a><a href='#L16'>16</a>
<a name='L17'></a><a href='#L17'>17</a>
<a name='L18'></a><a href='#L18'>18</a>
<a name='L19'></a><a href='#L19'>19</a>
<a name='L20'></a><a href='#L20'>20</a>
<a name='L21'></a><a href='#L21'>21</a>
<a name='L22'></a><a href='#L22'>22</a>
<a name='L23'></a><a href='#L23'>23</a>
<a name='L24'></a><a href='#L24'>24</a>
<a name='L25'></a><a href='#L25'>25</a>
<a name='L26'></a><a href='#L26'>26</a>
<a name='L27'></a><a href='#L27'>27</a>
<a name='L28'></a><a href='#L28'>28</a>
<a name='L29'></a><a href='#L29'>29</a>
<a name='L30'></a><a href='#L30'>30</a>
<a name='L31'></a><a href='#L31'>31</a>
<a name='L32'></a><a href='#L32'>32</a>
<a name='L33'></a><a href='#L33'>33</a>
<a name='L34'></a><a href='#L34'>34</a>
<a name='L35'></a><a href='#L35'>35</a>
<a name='L36'></a><a href='#L36'>36</a>
<a name='L37'></a><a href='#L37'>37</a>
<a name='L38'></a><a href='#L38'>38</a>
<a name='L39'></a><a href='#L39'>39</a>
<a name='L40'></a><a href='#L40'>40</a>
<a name='L41'></a><a href='#L41'>41</a>
<a name='L42'></a><a href='#L42'>42</a>
<a name='L43'></a><a href='#L43'>43</a>
<a name='L44'></a><a href='#L44'>44</a>
<a name='L45'></a><a href='#L45'>45</a>
<a name='L46'></a><a href='#L46'>46</a>
<a name='L47'></a><a href='#L47'>47</a>
<a name='L48'></a><a href='#L48'>48</a>
<a name='L49'></a><a href='#L49'>49</a>
<a name='L50'></a><a href='#L50'>50</a>
<a name='L51'></a><a href='#L51'>51</a>
<a name='L52'></a><a href='#L52'>52</a>
<a name='L53'></a><a href='#L53'>53</a>
<a name='L54'></a><a href='#L54'>54</a>
<a name='L55'></a><a href='#L55'>55</a>
<a name='L56'></a><a href='#L56'>56</a>
<a name='L57'></a><a href='#L57'>57</a>
<a name='L58'></a><a href='#L58'>58</a>
<a name='L59'></a><a href='#L59'>59</a>
<a name='L60'></a><a href='#L60'>60</a>
<a name='L61'></a><a href='#L61'>61</a>
<a name='L62'></a><a href='#L62'>62</a>
<a name='L63'></a><a href='#L63'>63</a>
<a name='L64'></a><a href='#L64'>64</a>
<a name='L65'></a><a href='#L65'>65</a>
<a name='L66'></a><a href='#L66'>66</a>
<a name='L67'></a><a href='#L67'>67</a>
<a name='L68'></a><a href='#L68'>68</a>
<a name='L69'></a><a href='#L69'>69</a>
<a name='L70'></a><a href='#L70'>70</a>
<a name='L71'></a><a href='#L71'>71</a>
<a name='L72'></a><a href='#L72'>72</a>
<a name='L73'></a><a href='#L73'>73</a>
<a name='L74'></a><a href='#L74'>74</a>
<a name='L75'></a><a href='#L75'>75</a>
<a name='L76'></a><a href='#L76'>76</a>
<a name='L77'></a><a href='#L77'>77</a>
<a name='L78'></a><a href='#L78'>78</a>
<a name='L79'></a><a href='#L79'>79</a>
<a name='L80'></a><a href='#L80'>80</a>
<a name='L81'></a><a href='#L81'>81</a>
<a name='L82'></a><a href='#L82'>82</a>
<a name='L83'></a><a href='#L83'>83</a>
<a name='L84'></a><a href='#L84'>84</a>
<a name='L85'></a><a href='#L85'>85</a>
<a name='L86'></a><a href='#L86'>86</a>
<a name='L87'></a><a href='#L87'>87</a>
<a name='L88'></a><a href='#L88'>88</a>
<a name='L89'></a><a href='#L89'>89</a>
<a name='L90'></a><a href='#L90'>90</a>
<a name='L91'></a><a href='#L91'>91</a>
<a name='L92'></a><a href='#L92'>92</a>
<a name='L93'></a><a href='#L93'>93</a>
<a name='L94'></a><a href='#L94'>94</a>
<a name='L95'></a><a href='#L95'>95</a>
<a name='L96'></a><a href='#L96'>96</a>
<a name='L97'></a><a href='#L97'>97</a>
<a name='L98'></a><a href='#L98'>98</a>
<a name='L99'></a><a href='#L99'>99</a>
<a name='L100'></a><a href='#L100'>100</a>
<a name='L101'></a><a href='#L101'>101</a>
<a name='L102'></a><a href='#L102'>102</a>
<a name='L103'></a><a href='#L103'>103</a>
<a name='L104'></a><a href='#L104'>104</a>
<a name='L105'></a><a href='#L105'>105</a>
<a name='L106'></a><a href='#L106'>106</a>
<a name='L107'></a><a href='#L107'>107</a>
<a name='L108'></a><a href='#L108'>108</a>
<a name='L109'></a><a href='#L109'>109</a>
<a name='L110'></a><a href='#L110'>110</a>
<a name='L111'></a><a href='#L111'>111</a>
<a name='L112'></a><a href='#L112'>112</a>
<a name='L113'></a><a href='#L113'>113</a>
<a name='L114'></a><a href='#L114'>114</a>
<a name='L115'></a><a href='#L115'>115</a>
<a name='L116'></a><a href='#L116'>116</a>
<a name='L117'></a><a href='#L117'>117</a>
<a name='L118'></a><a href='#L118'>118</a>
<a name='L119'></a><a href='#L119'>119</a>
<a name='L120'></a><a href='#L120'>120</a>
<a name='L121'></a><a href='#L121'>121</a>
<a name='L122'></a><a href='#L122'>122</a>
<a name='L123'></a><a href='#L123'>123</a>
<a name='L124'></a><a href='#L124'>124</a>
<a name='L125'></a><a href='#L125'>125</a>
<a name='L126'></a><a href='#L126'>126</a>
<a name='L127'></a><a href='#L127'>127</a>
<a name='L128'></a><a href='#L128'>128</a>
<a name='L129'></a><a href='#L129'>129</a>
<a name='L130'></a><a href='#L130'>130</a>
<a name='L131'></a><a href='#L131'>131</a>
<a name='L132'></a><a href='#L132'>132</a>
<a name='L133'></a><a href='#L133'>133</a>
<a name='L134'></a><a href='#L134'>134</a>
<a name='L135'></a><a href='#L135'>135</a></td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
&nbsp;
import InputField from '../misc/InputField';
import Select from '../misc/Select';
import RadioButtons from '../misc/RadioButtons';
import DatePicker from '../misc/DatePicker';
import Reservation from '../../types/Reservation';
import Store from '../../types/Store';
import Status from '../../types/Status';
import { isFromTodayOn, cleanObject, formatDate } from '../../utils/utils';
&nbsp;
import styles from '../../styles/UpsertReservationForm.module.css';
&nbsp;
// Create a new reservation, or Update an existing reservation
const UpsertReservationForm = ({
  storesMap,
  statusesMap,
  formId,
  onSubmit,
  onClose,
  // below prop is for UPDATING an existing entry; if not supplied, the user is creating an entry
  reservation: existingReservation,
}) =&gt; {
  const [name, setName] = useState(existingReservation?.name || '');
  const [storeName, setStoreName] = useState(storesMap[existingReservation?.storeId]?.name || '');
  const [statusName, setStatusName] = useState(statusesMap[existingReservation?.statusId]?.name || '');
  const [date, setDate] = useState(existingReservation?.date ? new Date(existingReservation.date) : null);
&nbsp;
  const ref = useRef(null);
&nbsp;
  const storeOptions = Object.values(storesMap);
  const statusNames = Object.values(statusesMap).map((status) =&gt; status.name);
&nbsp;
  const handleSubmit = <span class="fstat-no" title="function not covered" >(e</span>) =&gt; {
<span class="cstat-no" title="statement not covered" >    e.preventDefault();</span>
&nbsp;
<span class="cstat-no" title="statement not covered" >    if (!date) {</span>
<span class="cstat-no" title="statement not covered" >      ref.current.setOpen(true);</span>
    } else <span class="cstat-no" title="statement not covered" >if (name &amp;&amp; storeName &amp;&amp; statusName) {</span>
      const selectedStore = <span class="cstat-no" title="statement not covered" >Object.values(storesMap).find(<span class="fstat-no" title="function not covered" >(s</span>tore) =&gt; <span class="cstat-no" title="statement not covered" >store.name === storeName)</span>;</span>
      const selectedStatus = <span class="cstat-no" title="statement not covered" >Object.values(statusesMap).find(<span class="fstat-no" title="function not covered" >(s</span>tatus) =&gt; <span class="cstat-no" title="statement not covered" >status.name === statusName)</span>;</span>
<span class="cstat-no" title="statement not covered" >      if (selectedStore &amp;&amp; selectedStatus) {</span>
        const reservationToUpsert = <span class="cstat-no" title="statement not covered" >{</span>
          id: existingReservation?.id, // present only when updating an existing reservation
          name,
          storeId: selectedStore?.id,
          statusId: selectedStatus?.id,
          date: formatDate(date),
        };
&nbsp;
        // if it is in UPDATE mode but no change is made, no action should be dispatched
        let allKeysUnchanged;
<span class="cstat-no" title="statement not covered" >        if (existingReservation</span>
          &amp;&amp; existingReservation.name == name
          &amp;&amp; existingReservation.storeId === selectedStore?.id
          &amp;&amp; existingReservation.statusId === selectedStatus?.id
          &amp;&amp; formatDate(existingReservation.date) === formatDate(date)
        ) {
<span class="cstat-no" title="statement not covered" >          allKeysUnchanged = true;</span>
        }
&nbsp;
<span class="cstat-no" title="statement not covered" >        if (!allKeysUnchanged) {</span>
<span class="cstat-no" title="statement not covered" >          onSubmit(cleanObject(reservationToUpsert)); </span>// wipe out undefined / empty props
<span class="cstat-no" title="statement not covered" >          e.preventDefault();</span>
        } else {
<span class="cstat-no" title="statement not covered" >          onClose();</span>
        }
      }
    } else {
<span class="cstat-no" title="statement not covered" >      throw new Error('Store/Status not exist - please refresh page and retry.');</span>
    }
  };
&nbsp;
  return (
    &lt;form
      id={formId}
      className={styles.upsertReservationModalForm}
      onSubmit={handleSubmit}
    &gt;
      &lt;InputField
        label='Name'
        name='name'
        value={name}
        onChange={setName}
        required
      /&gt;
&nbsp;
      &lt;Select
        options={storeOptions}
        onChange={setStoreName}
        label='Store'
        name='status'
        value={storeName}
        prependBlankOption
        required
      /&gt;
&nbsp;
      &lt;RadioButtons
        values={statusNames}
        onChange={setStatusName}
        label='Status'
        name='status'
        value={statusName}
        prependBlankOption
        required
      /&gt;
&nbsp;
      &lt;DatePicker
        ref={ref}
        label='Date'
        required
        dateFormat="yyyy-MM-dd"
        selected={date}
        onChange={<span class="fstat-no" title="function not covered" >(n</span>ewDate) =&gt; <span class="cstat-no" title="statement not covered" >setDate(newDate)}</span>
        filterDate={isFromTodayOn}
      /&gt;
    &lt;/form&gt;
  );
};
&nbsp;
// *********************************************************************************************************************
// PROPS
// *********************************************************************************************************************
UpsertReservationForm.propTypes = {
  storesMap: PropTypes.objectOf(PropTypes.exact(Store)).isRequired,
  statusesMap: PropTypes.objectOf(PropTypes.exact(Status)).isRequired,
  formId: PropTypes.string,
  onSubmit: PropTypes.func,
  onClose: PropTypes.func,
  reservation: PropTypes.exact(Reservation),
};
&nbsp;
export default UpsertReservationForm;
&nbsp;</pre></td></tr></table></pre>

                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at Mon Jan 31 2022 15:37:22 GMT+0100 (Central European Standard Time)
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    