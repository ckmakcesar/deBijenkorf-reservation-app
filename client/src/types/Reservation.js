import * as PropTypes from 'prop-types';

const Reservation = {
  id: PropTypes.number.isRequired,
  storeId: PropTypes.number.isRequired,
  statusId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  time: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
};

export default Reservation;
