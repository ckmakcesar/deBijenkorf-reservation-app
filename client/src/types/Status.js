import * as PropTypes from 'prop-types';

const Status = {
  id: PropTypes.number.isRequired,
  code: PropTypes.string.isRequired,
};

export default Status;
