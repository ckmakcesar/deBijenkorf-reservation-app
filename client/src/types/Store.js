import * as PropTypes from 'prop-types';

const Store = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  address: PropTypes.string.isRequired,
};

export default Store;
