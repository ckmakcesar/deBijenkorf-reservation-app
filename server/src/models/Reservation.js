const { Sequelize, DataTypes } = require('sequelize');

const { sequelize } = require('.');

const Reservation = sequelize.define('reservation', {
  id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
  name: { type: DataTypes.STRING, allowNull: false },
  storeId: { type: DataTypes.INTEGER, allowNull: false },
  statusId: { type: DataTypes.INTEGER, allowNull: false },
  time: { type: DataTypes.DATEONLY, allowNull: false },
  createdAt: { type: DataTypes.DATE, allowNull: false, defaultValue: Sequelize.NOW },
}, {
  freezeTableName: true,
  underscored: true,
  version: false,
  updatedAt: false,
});

module.exports = Reservation;
