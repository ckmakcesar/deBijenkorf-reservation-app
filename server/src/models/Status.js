const { DataTypes } = require('sequelize');

const { sequelize } = require('.');

const Status = sequelize.define('status', {
  id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
  code: { type: DataTypes.STRING, allowNull: false, unique: true },
  name: { type: DataTypes.STRING, allowNull: false, unique: true },
}, {
  freezeTableName: true,
  underscored: true,
  version: false,
  createdAt: false,
  updatedAt: false,
});

module.exports = Status;
