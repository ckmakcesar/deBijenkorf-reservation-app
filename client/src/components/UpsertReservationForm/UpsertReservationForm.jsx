import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';

import InputField from '../misc/InputField';
import Select from '../misc/Select';
import RadioButtons from '../misc/RadioButtons';
import DatePicker from '../misc/DatePicker';
import Reservation from '../../types/Reservation';
import Store from '../../types/Store';
import Status from '../../types/Status';
import { isFromTodayOn, cleanObject, formatDate } from '../../utils/utils';

import styles from '../../styles/UpsertReservationForm.module.css';

const UpsertReservationForm = ({
  storesMap,
  statusesMap,
  formId,
  onSubmit,
  // below prop is for editing an existing entry; if not supplied, the user is creating an entry
  reservation: existingReservation,
}) => {
  const [name, setName] = useState(existingReservation?.name || '');
  const [storeName, setStoreName] = useState(storesMap[existingReservation?.storeId]?.name || '');
  const [statusName, setStatusName] = useState(statusesMap[existingReservation?.statusId]?.name || '');
  const [date, setDate] = useState(existingReservation?.date ? new Date(existingReservation.date) : null);

  const ref = useRef(null);

  const storeOptions = Object.values(storesMap);
  const statusNames = Object.values(statusesMap).map((status) => status.name);

  const handleSubmit = (e) => {
    e.preventDefault();

    if (!date) {
      ref.current.setOpen(true);
    } else if (name && storeName && statusName) {
      const selectedStore = Object.values(storesMap).find((store) => store.name === storeName);
      const selectedStatus = Object.values(statusesMap).find((status) => status.name === statusName);
      if (selectedStore && selectedStatus) {
        const reservationToUpsert = {
          id: existingReservation?.id, // present only when updating an existing reservation
          name,
          storeId: selectedStore?.id,
          statusId: selectedStatus?.id,
          date: formatDate(date),
        };
        onSubmit(cleanObject(reservationToUpsert)); // wipe out undefined / empty props
        e.preventDefault();
      }
    } else {
      throw new Error('Store/Status not exist - please refresh page and retry.');
    }
  };

  return (
    <form
      id={formId}
      className={styles.upsertReservationModalForm}
      onSubmit={handleSubmit}
    >
      <InputField
        label='Name'
        name='name'
        value={name}
        onChange={setName}
        required
      />

      <Select
        options={storeOptions}
        onChange={setStoreName}
        label='Store'
        name='status'
        value={storeName}
        prependBlankOption
        required
      />

      <RadioButtons
        values={statusNames}
        onChange={setStatusName}
        label='Status'
        name='status'
        value={statusName}
        prependBlankOption
        required
      />

      <DatePicker
        ref={ref}
        label='Date'
        required
        dateFormat="yyyy-MM-dd"
        selected={date}
        onChange={(newDate) => setDate(newDate)}
        filterDate={isFromTodayOn}
      />
    </form>
  );
};

// *********************************************************************************************************************
// PROPS
// *********************************************************************************************************************
UpsertReservationForm.propTypes = {
  storesMap: PropTypes.objectOf(PropTypes.exact(Store)).isRequired,
  statusesMap: PropTypes.objectOf(PropTypes.exact(Status)).isRequired,
  formId: PropTypes.string,
  onSubmit: PropTypes.func,
  reservation: PropTypes.exact(Reservation),
};

export default UpsertReservationForm;
