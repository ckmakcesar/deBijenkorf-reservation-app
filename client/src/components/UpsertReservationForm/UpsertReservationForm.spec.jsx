import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import UpsertReservationForm from './UpsertReservationForm';
import { mockReservations } from '../../tests/data/Reservation';
import { mockStoresMap } from '../../tests/data/Store';
import { mockStatusesMap } from '../../tests/data/Status';

describe('component:UpsertReservationForm', () => {
  it('Should render UpsertReservationForm in the CREATE RESERVATION mode', () => {
    const wrapper = shallow(
      <UpsertReservationForm
        storesMap={mockStoresMap}
        statusesMap={mockStatusesMap}
        formId='form-101'
        onSubmit={jest.fn()}
        onClose={jest.fn()}
      />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('Should render UpsertReservationForm in the UPDATE RESERVATION mode', () => {
    const wrapper = shallow(
      <UpsertReservationForm
        storesMap={mockStoresMap}
        statusesMap={mockStatusesMap}
        formId='form-202'
        onSubmit={jest.fn()}
        onClose={jest.fn()}
        reservation={mockReservations[0]}
      />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
