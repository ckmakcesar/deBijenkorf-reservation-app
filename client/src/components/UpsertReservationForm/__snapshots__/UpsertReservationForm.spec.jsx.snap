// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:UpsertReservationForm Should render UpsertReservationForm in the CREATE RESERVATION mode 1`] = `
<form
  className="upsertReservationModalForm"
  id="form-101"
  onSubmit={[Function]}
>
  <InputField
    disabled={false}
    label="Name"
    max={NaN}
    min={NaN}
    name="name"
    onChange={[Function]}
    required={true}
    type=""
    value=""
  />
  <Select
    disabled={false}
    label="Store"
    name="status"
    onChange={[Function]}
    options={
      Array [
        Object {
          "address": "Mid Amsterdam",
          "id": 1,
          "name": "Centraal",
        },
        Object {
          "address": "Nord Oost Amsterdam",
          "id": 2,
          "name": "Zeeburg",
        },
        Object {
          "address": "Oost Amsterdam",
          "id": 3,
          "name": "Diemen",
        },
        Object {
          "address": "Zuid Amsterdam",
          "id": 4,
          "name": "Amstelveen",
        },
        Object {
          "address": "West Amsterdam",
          "id": 5,
          "name": "Sloten",
        },
      ]
    }
    prependBlankOption={true}
    required={true}
    value=""
  />
  <RadioButtons
    disabled={false}
    label="Status"
    name="status"
    onChange={[Function]}
    prependBlankOption={true}
    required={true}
    value=""
    values={
      Array [
        "Ready",
        "In Progress",
        "Todo",
      ]
    }
  />
  <DatePicker
    dateFormat="yyyy-MM-dd"
    filterDate={[Function]}
    label="Date"
    onChange={[Function]}
    required={true}
    selected={null}
  />
</form>
`;

exports[`component:UpsertReservationForm Should render UpsertReservationForm in the UPDATE RESERVATION mode 1`] = `
<form
  className="upsertReservationModalForm"
  id="form-202"
  onSubmit={[Function]}
>
  <InputField
    disabled={false}
    label="Name"
    max={NaN}
    min={NaN}
    name="name"
    onChange={[Function]}
    required={true}
    type=""
    value="Repair"
  />
  <Select
    disabled={false}
    label="Store"
    name="status"
    onChange={[Function]}
    options={
      Array [
        Object {
          "address": "Mid Amsterdam",
          "id": 1,
          "name": "Centraal",
        },
        Object {
          "address": "Nord Oost Amsterdam",
          "id": 2,
          "name": "Zeeburg",
        },
        Object {
          "address": "Oost Amsterdam",
          "id": 3,
          "name": "Diemen",
        },
        Object {
          "address": "Zuid Amsterdam",
          "id": 4,
          "name": "Amstelveen",
        },
        Object {
          "address": "West Amsterdam",
          "id": 5,
          "name": "Sloten",
        },
      ]
    }
    prependBlankOption={true}
    required={true}
    value="Centraal"
  />
  <RadioButtons
    disabled={false}
    label="Status"
    name="status"
    onChange={[Function]}
    prependBlankOption={true}
    required={true}
    value="Todo"
    values={
      Array [
        "Ready",
        "In Progress",
        "Todo",
      ]
    }
  />
  <DatePicker
    dateFormat="yyyy-MM-dd"
    filterDate={[Function]}
    label="Date"
    onChange={[Function]}
    required={true}
    selected={2022-01-26T23:00:00.000Z}
  />
</form>
`;
